### PR TITLE
Issues/1273 shacl benchmark

### PR DESCRIPTION
**Pull request by hmottestad**
This PR addresses GitHub issue: eclipse/rdf4j#1273 .

Briefly describe the changes proposed in this PR:

* made benchmarks run again
* minor adjustments for performance
* only persist memory store used for storing shacl rules if the base sail is persisted

### CODE REVIEW

File: shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java [hunk 1]
```diff
@@ -217,42 +217,37 @@ public void initialize() throws SailException {
...
-			}
-			catch (IOException e) {
-				throw new SailConfigException(e);
+			String path = super.getBaseSail().getDataDir().getPath();
+			if (path.endsWith("/")) {
+				path = path.substring(0, path.length() - 1);
 			}
+			path = path + "-shapes-graph/";
+
+			shapesRepo = new SailRepository(new MemoryStore(new File(path)));
```
Comment (hmottestad) [inline shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java]: @jeenbroekstra I changed the code you wrote for persisting the shacl rules memory store. Not it only persists if the base sail persists. It's for two reasons.
 - performance
 - abide by the users intended use and not force writing to disk

File: shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSailConnection.java [hunk 2]
```diff
@@ -337,25 +308,47 @@ public void prepare() throws SailException {
...
+		refreshShapes(shapesConnection);
 
-			if (!sail.isIgnoreNoShapesLoadedException()
-				&& ((!addedStatementsSet.isEmpty() || !removedStatementsSet.isEmpty())
-				&& sail.getNodeShapes().isEmpty())) {
-				throw new NoShapesLoadedException();
-			}
+		if (!sail.isIgnoreNoShapesLoadedException()
+			&& ((!addedStatementsSet.isEmpty() || !removedStatementsSet.isEmpty())
+			&& sail.getNodeShapes().isEmpty())) {
```
Comment (jeenbroekstra) [inline shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSailConnection.java]: Aren't you introducing a possible race condition here by removing the synchronized block around this? What if a parallel connection also modified the shapes data?

File: shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java [hunk 3]
```diff
@@ -217,42 +217,37 @@ public void initialize() throws SailException {
...
 		if (super.getBaseSail().getDataDir() != null) {
-			path = super.getBaseSail().getDataDir().getPath();
-		}
-		else {
-			try {
-				path = Files.createTempDirectory("shacl-shapes").toString();
-			}
-			catch (IOException e) {
-				throw new SailConfigException(e);
+			String path = super.getBaseSail().getDataDir().getPath();
```
[the remark this answers is not in the document]
Comment (hmottestad) [reply shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSail.java]: It's a bit hard to read in the diff. Code looks like this:

```java
if (super.getBaseSail().getDataDir() != null) {
	String path = super.getBaseSail().getDataDir().getPath();
	if (path.endsWith("/")) {
		path = path.substring(0, path.length() - 1);
	}
	path = path + "-shapes-graph/";

	shapesRepo = new SailRepository(new MemoryStore(new File(path)));
} else {
	shapesRepo = new SailRepository(new MemoryStore());
}
```

File: shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSailConnection.java (hunk 2 above)
Comment (hmottestad) [reply shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSailConnection.java]: There is a lot of issues with race conditions with the shapes being cached in the ShaclSail and not in the connection. I'll make a new issue to track the work to for that.

**Pull request merged**